Implement option to output information as OpenMetrics time series

lelutin opened the pull request:
This new `-o` option will make needrestart output information in a format that can be scraped by Prometheus or any other daemon that ingests OpenMetrics format.

The -l, -w and -k options can be used in combination with -o in order to choose what information gets exported.

(Closes: #291)

lelutin commented:
note, I've chosen to split the version information out into a dedicated `...build_info` time series. This makes it possible to graph the timestamp normally without risking discontinuity if needrestart and/or perl is upgraded and thus the labels change, and the dedicated metric can be used just for aggregation if needs be.

anarcat reviewed (changes requested):
good job untangling that code, it's hard to find your way in there, and you seem to have found the right spots!

but the metrics exposition formats need a little work. i think once that's done, it's ready to merge, as far as i'm concerned.

it would be helpful to show sample outputs, at least in the commit log, that show the previous nagios and plain batch mode still works as well... the openmetrics output could be useful in the README or manpage directly as well.

lelutin commented on needrestart:
```diff
@@ -1397,6 +1411,37 @@ if($opt_p) {
+
+    my $ometric_now = time();
+    print "# HELP needrestart_timestamp when, in unix timestamp, needrestart was last updated\n";
+    print "# TYPE needrestart_timestamp gauge\n";
```
oh interesting I should take a better look at the prometheus documentation about writing exporters, there seems to be a couple of good hints in there that I haven't yet integrated.

now that you mention it, it seems indeed to be redundant.. the timing info will already be there in the time series. so let's drop that metric. I'll push this change in a couple minutes

lelutin commented on needrestart:
```diff
@@ -1397,6 +1411,37 @@ if($opt_p) {
+    if ($opt_k) {
+        print "# HELP needrestart_kernel_info information about the kernel\n";
+        print "# TYPE needrestart_kernel_info info\n";
+        print "needrestart_kernel_info{running=$ometric_kernel_values{krunning},expected=$ometric_kernel_values{kexpected}} $ometric_kernel_values{kstatus}\n";
```
no currently it will look like `needrestart_kernel_info{running=6.9.7,expected=6.9.7} current` on success

and `needrestart_kernel_info{running=6.9.6,expected=6.9.7} obsolete` on failure.


that detail was something that I was really not sure about how to output.. outputting a binary value could be interesting in the case of the kernel since there's currently only two possible states. however, for microcode, there are three different states, `unknown`, `current` and `obsolete`. so to keep things consistent, I decided to output the state "name" for both.

should the kernel_info metric be a bool instead?

lelutin commented on needrestart:
```diff
@@ -1397,6 +1411,37 @@ if($opt_p) {
+        my $ometric_ucode_status = ("unknown", "current", "obsolete")[$ucode_result];
+        print "# HELP needrestart_ucode_info information about the CPU microcode\n";
+        print "# TYPE needrestart_ucode_info info\n";
+        print "needrestart_ucode_info{running=$ometric_ucode_current,expected=$ometric_ucode_expected} $ometric_ucode_status\n";
```
oh wow, I should have read this comment before answering above. ...I now understand that in my reading of that openmetrics document before sending this PR I did *not* understand what it was trying to communicate.. why is it structured in reverse like that? blah I mean if they had presented the ABNF form *before* explaining what each part consists of it would have made things so much less confusing.


ok well I'll fix the values then. I'll have to figure how to represent the states for microcode though. I could conflate `unknown` and `obsolete` into the boolean for failed (e.g. 0), but then we're losing information in that case. It could be an index (so in the codebase, simply the value of ucode_result).. I'll think about this for some time and send a change soon

lelutin commented:
I've reworked this branch to incorporate feedback from @anarcat and other details found while reading the OpenMetrics spec more in depth and I've just force-pushed the result.

what's to be expected as being changed in this force-push;

* the `-l` option in combination with `-o` now requires root access, otherwise the metric for outdated libraries was always set to 0 since an unprivileged user can't dig around all running processes to find out this information
* the code and metrics now talk about "outdated libraries" and "processes with outdated libraries", which should be more precise for what we're exposing
* the metrics were renamed to be more descriptive of what they expose
* the man page now has some description of which metrics will be exposed and their types
* the timestamp metric was removed
* the info-type metrics for kernel and microcode, which were wrongly typed and implemented, now use the StateSet type to expose all of the possible states
* there's an additional info-type metric for kernel and microcode in order to expose the current and expected versions. This is instead of attaching those versions as labels to all entries in the StateSet metrics. it reduces useless duplication and still makes it possible to group by those metrics if needed in order to filter or add those labels to results
* the metadata for all MetricFamilies were reordered as per the OpenMetric recommendation (type, then help -- unit was not used)
* a trailing `# EOF\n` was added to properly end the `exposition` as defined by OpenMetrics

liske commented:
Thanks!

anarcat commented:
awesome, thanks for the fixes
